File: jiuzhang/word-search-ii/word-search-ii.cpp
/**
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
 */

struct Node {
    Node * ch[26];
    string str;
    Node() {
        for (int i = 0; i < 26; ++i)
            ch[i] = NULL;
            str = "";
    }
};

class Solution {
public:
    vector results;
    Node * root;
    void insert(Node* p, string s) {
        int len = s.size();
        for (int i = 0; i < len; ++i) {
            if (p->ch[s[i]-"a"] == NULL)
                p->ch[s[i]-"a"] = new Node();
            p = p->ch[s[i]-"a"];
        }   
        p->str = s;
    }
    void search(vector > &board, vector<vector<bool> > &mask, Node* p, int x, int y) {
        if (p->str != "") {
            results.push_back(p->str);
            p->str = "";
        }

        mask[x][y] = false;
        if (y + 1 ch[board[x][y+1]-"a"] != NULL) {
           search(board,mask, p->ch[board[x][y+1]-"a"], x, y+1);
        }

        if (x + 1 ch[board[x+1][y]-"a"] != NULL) {
           search(board,mask, p->ch[board[x+1][y]-"a"], x+1, y);
        }

        if (y - 1 >= 0 && mask[x][y-1] && p->ch[board[x][y-1]-"a"] != NULL) {
           search(board,mask, p->ch[board[x][y-1]-"a"], x, y-1);
        }

        if (x - 1 >= 0 && mask[x-1][y] && p->ch[board[x-1][y]-"a"] != NULL) {
           search(board,mask, p->ch[board[x-1][y]-"a"], x-1, y);
        }
        mask[x][y] = true;
    }
    /**
     * @param board: A list of lists of character
     * @param words: A list of string
     * @return: A list of string
     */
    vector wordSearchII(vector<vector<char> > &board, vector<string> &words) {
        // write your code here
        root = new Node();
        int len = words.size();
        for (int i = 0; i < len; ++i) { 
            insert(root, words[i]);
        }
        vector > mask(board.size(), vector<bool>(board[0].size(), true));
        if (board.size() < 1) return results;
        for (int i = 0; i <board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
               if (root->ch[board[i][j]-"a"] != NULL) {
                    search(board, mask, root->ch[board[i][j]-"a"], i, j);
                }
            }
        }
        return results;
    }
};

File: jiuzhang/word-search-ii/word-search-ii.py
"""
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
"""

class Trie:
    
    def __init__(self):
        self.children = {}
        self.flag = False
        self.hasWord = False
    
    def put(self, key):
        if key == "":
            self.flag = True
            self.hasWord = True
            return
        
        if key[0] not in self.children:
            self.children[key[0]] = Trie()
        self.children[key[0]].put(key[1:])
        self.hasWord = True
    
    def pop(self, key):
        if key == "":
            self.flag = False
            self.hasWord = False
            return
        if key[0] not in self.children:
            return
        self.children[key[0]].pop(key[1:])
        self.hasWord = any([child.hasWord for child in self.children.values()])
        
    def has(self, key):
        if key == "":
            return self.flag
        
        if not self.hasWord:
            return False
        if key[0] not in self.children:
            return False
        return self.children[key[0]].has(key[1:])


class Solution:
    DIRECT_X = [1, 0, 0, -1]
    DIRECT_Y = [0, 1, -1, 0]
    # @param board, a list of lists of 1 length string
    # @param words: A list of string
    # @return: A list of string
    def wordSearchII(self, board, words):
        trie = Trie()
        for word in words:
            trie.put(word)
        
        self.results = {}
        for r in range(len(board)):
            for c in range(len(board[0])):
                self.search(trie, trie, board, r, c, [])
        return self.results.keys()
                
    def search(self, root, trie, board, x, y, chars):
        char = board[x][y]
        if char not in trie.children:
            return
        chars.append(char)
        trie = trie.children[char]
        if trie.flag:
            self.results["".join(chars)] = True
            root.pop("".join(chars))
        
        board[x][y] = "."
        for i in range(4):
            r = x + self.DIRECT_X[i]
            c = y + self.DIRECT_Y[i]
            if r < 0 or r == len(board) or c < 0 or c == len(board[0]):
                continue
            self.search(root, trie, board, r, c, chars)
        board[x][y] = char
        chars.pop()
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
